### PR TITLE
Feature/integration tests

### PR DESCRIPTION
**Pull request by Rana-Mudassir**
# About this PR
 
- [x] Fixed the N + 1 problem(See the logs in the screenshots below).
- [x]  Wrote tests for all my views.

- Log before N + 1 fix
beforen+!

- Log After N + 1 fix
aftern+1

Thanks for Reviewing 😀

### CODE REVIEW

**Review by Ismailco — CHANGES_REQUESTED** [review]
### Hi @Rana-Mudassir, @FrankMugagga, 

Good job so far 👍🏼 
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights

✔️ No linter errors
✔️ Used Gitflow
✔️ Work documented professionally

## Required Changes ♻️
- [x] Greate work however I can't find any test for capybara, all the tests are for model functions and validations. Kindly make sure to write integration tests for each view in your project: 

Screenshot_20230714_024905


- [x] Kindly make sure to fix all the tests so they pass, as for now it throws an error 👇🏼 
Screenshot_20230714_023151

_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by Rana-Mudassir** [comment]
Hi, Fixed Errors and Now all tests working. 😀

File: spec/views/users/show_spec.rb [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    end
+
+    it 'displays the number of posts for a user' do
+      expect(page).to have_content('Number of posts: 4')
+    end
+
+    it 'displays the user bio' do
+      expect(page).to have_content(user.bio)
+    end
+
```
Comment (Kamba56) [inline spec/views/users/show_spec.rb:61]: - [x] Great job adding rspec tests for users show page. However, there are a few tests missing. As demanded by project requirements, kindly add the following tests.
  - Add a test to check if show page displays three most recent posts by the user
  - A test to check if clicking on the see all posts button will redirect to the posts index page
  - A test that checks the page to redirect to post show page when any of the three recent posts is clicked.

File: spec/views/posts/index_spec.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+                  comments: [])
+    ]
+  end
+
+  it 'displays the user profile information and posts' do
+    visit user_posts_path(@user)
+
+    expect(page).to have_selector('h2', text: @user.name)
+    expect(page).to have_selector('img[src="https://media.istockphoto.com/id/1289461328/photo/portrait-of-a-handsome-black-man.jpg?s=612x612&w=0&k=20&c=y_mzB0Tbe5LErNy6pqfY7sz2HiDT7fOAUCwupN3-Bg4="]')
+
```
Comment (Kamba56) [inline spec/views/posts/index_spec.rb:22]: - [x] Your posts index test is missing some few test cases. kindly add the following test cases as demanded by project requirements.
  - The page displays number of posts the user has written
  - The page should display part of the post body
  - Test whether clicking on a post redirects user to the post showpage

File: spec/views/posts/show_spec.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  end
+
+  it 'displays the post details and comments' do
+    within('.post-details') do
+      expect(page).to have_selector('.user-count span', text: 'Comments: 1')
+      expect(page).to have_selector('.user-count span', text: 'Likes: 0')
+      expect(page).to have_selector('h2', text: @post.title)
+      expect(page).to have_content(@user.name)
+    end
+
```
Comment (Kamba56) [inline spec/views/posts/show_spec.rb:18]: - [x] You are missing a few test cases for the posts show page. Kindly add these as demanded by project requirements.
  - A test case to check if the post body is displayed
  - Check if each comment is display
  - Check if the username of a commenter for each comment is displayed

**Review by Kamba56 — CHANGES_REQUESTED** [review]
Hi @Rana-Mudassir , 

Great work on making the changes requested by a previous reviewer 👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights
- [x] Great work solving n+1 problem 👏👏👏
- [x] Great work adding test for pages👏👏👏

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/views/users/show_spec.rb (hunk 1 above)
Comment (Rana-Mudassir) [reply spec/views/users/show_spec.rb:61]: Hi, thanks for the review. will you elaborate me one thing, the tests you mentioned above are remaining only of users' show page?

File: spec/views/posts/index_spec.rb (hunk 2 above)
Comment (Rana-Mudassir) [reply spec/views/posts/index_spec.rb:22]: All tests added. Thanks

File: spec/views/posts/show_spec.rb (hunk 3 above)
Comment (Rana-Mudassir) [reply spec/views/posts/show_spec.rb:18]: All tests added. Thanks

File: spec/views/users/show_spec.rb (hunk 1 above)
Comment (Rana-Mudassir) [reply spec/views/users/show_spec.rb:61]: All tests added. Thanks

**Review by codepantha — APPROVED** [review]
# STATUS: APPROVED 🟢 ✔️ 

## Hi @Rana-Mudassir 👋, 

Great job implementing the required changes 👏 
Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights 🎯 
✔️ Descriptive PR
✔️ Implemented changes suggested by CR 📌 
✔️ All tests are green ✔️ 📌 

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**